### PR TITLE
Water - Kayla Johnson

### PR DESCRIPTION
**Pull request by Kaylaj89**
# Media Ranker
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
---      | ---
Describe a custom model method you wrote. | I had a lot of issues with scope this project and ended up writing my custom methods in the homepages controller (spotlight and top ten method).
Describe how you approached testing that model method. What edge cases did you come up with? | I only did a couple of validation tests. I struggled with this project--took much longer than anticipated and decided to focus on getting functionality figured out before testing.
What are `session` and `flash`? What is the difference between them? | A session tracks your movement throughout your session until you close your browser -- a flash is a brief message that pops up until your next page movement
What was one thing that you gained more clarity on through this assignment? | 
What is the Heroku URL of your deployed application? | Working on it (having a lot of issues trying to get it to work)

# Assignment Submission: Media Ranker

Congratulations! You're submitting your assignment. Please reflect on the assignment with these questions.

## Reflection

Prompt | Response
--- | ---
What was a custom model method you wrote? What was it responsible for doing? | 
Describe how you approached testing that model method. What edge cases did you come up with? | 
What are `session` and `flash`? What is the difference between them? | 
What was one thing that you gained more clarity on through this assignment? | 
What is the Heroku URL of your deployed application? |

### CODE REVIEW

File: test/controllers/users_controller_test.rb [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+      expect(session[:user_id]).must_equal user.id
+    end
+  end
+
+  describe "logout" do
+    it "can logout a logged in user" do
+      #Arrange
+      login()
+      expect[:session][:user_id].wont_be_nil
```
Comment (jmaddox19) [inline test/controllers/users_controller_test.rb:40]: Checkout the lesson on testing session to clarify how to test this properly :)

**Comment by jmaddox19** [comment]
# Media Ranker





## Functional Requirements: Manual Testing

| Criteria | yes/no |
| --- | --- |
| **Before logging in** | --
| 1. On index page, there are at most 10 pieces of media on three lists, and a Media Spotlight | homepage isn't working
| 2. Can go into a work's show page | ✔️ Yes, when I type the path manually (because homepage isn't working)
| 3. Verify unable to vote on a work, and get a flash message | ✔️
| 4. Can edit this work successfully, and get a flash message | ✔️
| 5. Can go to "View all media" page and see three lists of works, sorted by vote | ✔️
| 6. Verify unable to create a new work when the form is empty, and details about the validation errors are visible to the user through a flash message | Flash message shows up but no specific error messages
| 7. Can create a new work successfully. Note the URL for this work's show page | ✔️
| 8. Can delete this work successfully |✔️
| 9. Going back to the URL of this deleted work's show page produces a 404 or some redirect behavior (and does not try to produce a broken view) | ✔️
| 10. Verify that the "View all users" page lists no users | ✔️
| **Log in** | --
| 11. Logging in with a valid name changes the UI to "Logged in as" and "Logout" buttons | ✔️
| 12. Your username is listed in "View all users" page | ✔️
| 13. Verify that number of votes determines the Media Spotlight | Cannot view the media spotlight because homepage is broken
| 14. Voting on several different pieces of media affects the "Votes" tables shown in the work's show page and the user's show page | ✔️
| 15. Voting on the same work twice produces an error and flash message, and there is no extra vote | ✔️
| **Log out** | --
| 16. Logging out showed a flash message and changed the UI | ✔️
| 17. Logging in as a new user creates a new user | ✔️
| 18. Logging in as an already existing user has a specific flash message | ✔️

## Major Learning Goals/Code Review

| Criteria | yes/no |
| --- | --- |
| 1. Sees the full development cycle including deployment, and the app is deployed to Heroku | No but not required
| 2. Practices full-stack development and fulfilling story requirements: the styling, look, and feel of the app is similar to the original Media Ranker | ✔️ Looks really good!
| 3. Practices git with at least 25 small commits and meaningful commit messages | ✔️ So many commits with meaningful messages!

### Previous Rails learning, Building Complex Model Logic, DRYing up Rails Code

| Criteria | yes/no |
| --- | --- |
| 4. Routes file uses `resources` for works | ✔️
| 5. Routes file shows intention in limiting routes for voting, log-in functionality, and users | ✔️
| 6. The homepage view, all media view, and new works view use semantic HTML | The homepage view uses a lot of divs rather than section, article, etc. The other two views look good :)
| 7. The homepage view, all media view, and new works view use partials when appropriate | Yes for the form but there's a missed opportunity to use it for the homepage and all media views, which have some pieces in common. (Happy to talk more about that some time, if you have questions)
| 8. The model for media (likely named `work.rb`) `has_many` votes | ✔️
| 9. The model for media has methods to describe business logic, specifically for top ten and top media, possibly also for getting works by some category | 
| 10. Some controller, likely the `ApplicationController`, has a controller filter for finding a logged in user | ✔️
| 11. Some controller, likely the `WorksController`, has a controller filter for finding a work | ✔️
| 12. The `WorksController` uses strong params | ✔️
| 13. The `WorksController`'s code style is clean, and focused on working with requests, responses, `params`, `session`, `flash` | ✔️

### Testing Rails Apps

| Criteria | yes/no |
| --- | --- |
| 14. There are valid fixtures files used for users, votes, and works | works only
| 15. User model has tests with sections on validations (valid and invalid) and relationships (has votes) | 
| 16. Vote model has tests with sections on validations (valid and invalid) and relationships (belongs to a user, belongs to a vote) | 
| 17. Work model has tests with sections on validations (valid and invalid) and relationships (has votes) | none for relationships
| 18. Work model has tests with a section on all business logic methods in the model, including their edge cases | 

## Overall Feedback

Great work! You built a whole site users can log into, interact with, and vote to change the way things are ordered! That's a lot! I hope you've been able to celebrate this big win!

Technically the rubric would have me give you a red but the only big gap is a lack of testing, which I know you are focusing on for bEtsy and you accomplished all the other learning objectives so a Yellow feels much more appropriate.

| Overall Feedback | Criteria | yes/no |
| --- | --- | --- |
| Green (Meets/Exceeds Standards) | 14+ in Functional Requirements: Manual Testing && 14+ in Code Review | 
| Yellow (Approaches Standards) | 12+ in Functional Requirements: Manual Testing && 11+ in Code Review, or the instructor judges that this project needs special attention | ✔️
| Red (Not at Standard) | 0-10 in Code Review or 0-11 in Functional Reqs, or assignment is breaking/doesn’t run with less than 5 minutes of debugging, or the instructor judges that this project needs special attention | 





## Code Style Bonus Awards



Was the code particularly impressive in code style for any of these reasons (or more...?)

| Quality | Yes? |
| --- | --- |
| Perfect Indentation | ✅
| Elegant/Clever | ✅
| Descriptive/Readable | ✅
| Concise | ✅
| Logical/Organized | ✅
